---
layout: reveal_markdown
title: "Course overview, introduction to computational biology and bioconductor"
tags: slides 
date: 2021-12-09
---

## {{ page.title }}
---
## Course overview

- Course website: https://cphg.github.io/compbio/
- Quizzes: on "required" readings plus assignment topics.
- Programming assignments: please attempt on your own first


---
## Introduction to computational biology

What is computational biology?

<div class="fragment">bioinformatics?</div>

---
## Some history

- 1950-1960s Margaret Dayhoff
- 1970 First use of term bioinformatics (in dutch)
- 1974 "Biocomputing" gains popularity
- 1982 GenBank is founded
- 1985 Computer Applications in the Biosciences
- 1988 FASTA published
- 1988 NCBI is founded
- 1993 Intelligent Systems in Molecular Biology (ISMB) 
- 1994 Journal of Computational Biology

---
#### 1965 Atlas of protein sequence and structure
<img
	src="images/introduction/dayhoff1965toc.png"
	maxwidth="1100"
	style="background:white; padding: 10px; margin: 0px"><br>

---
#### 1965 Atlas of protein sequence and structure
<img
	src="images/introduction/dayhoff1965.png"
	maxwidth="1100"
	style="background:white; padding: 10px; margin: 0px"><br>

---
# 1969
<img
	src="images/introduction/dayhoff1969title.png"
	maxwidth="1100"
	style="background:white; padding: 10px; margin: 0px"><br>

---
# 1969
<img
	src="images/introduction/dayhoff1969alignment.png"
	width="1100"
	style="background:white; padding: 10px; margin: 0px"><br>

---
# 1969
<img
	src="images/introduction/dayhoff1969figure.png"
	width="1100"
	style="background:white; padding: 10px; margin: 0px"><br>

---
<img
	src="images/introduction/crovello1974.png"
	height="680"
	style="background:white; padding: 10px; margin: 0px"><br>

---
# 1974

Computers can be used for:

1. information retrieval
2. data accumulation
3. data analysis
4. systems analysis
5. simulate biological phenomena
6. control experiments
7. course review


---
# 1982

Genbank started by Walter Goad   
at Los Alamos National Laboratory. 

<https://www.ncbi.nlm.nih.gov/genbank/statistics/>


---
<img
	src="images/introduction/cabios1985.png"
	height="660"
	style="background:white; padding: 0px; margin: 0px"><br>

---
<img
	src="images/introduction/cabios1toc.png"
	height="660"
	style="background:white; padding: 0px; margin: 0px"><br>

---
<img
	src="images/introduction/beynon1985.png"
	height="680"
	style="background:white; padding: 10px; margin: 0px"><br>


---
<img
	src="images/introduction/cabios2toc.png"
	height="660"
	style="background:white; padding: 0px; margin: 0px"><br>

---
<img
	src="images/introduction/davis2023.png"
	height="660"
	style="background:white; padding: 0px; margin: 0px"><br>


---
<img
	src="images/introduction/pearson1988.png"
	height="660"
	style="background:white; padding: 0px; margin: 0px"><br>


---
# 1988

The NCBI was created by  
the Health Omnibus Extension Act P.L.100-607  
with $8 million.


---
<img
	src="images/introduction/ncbi.png"
	maxwidth="1100"
	style="background:white; padding: 0px; margin: 0px"><br>


---
### ISMB 1993

We spent a lot of time until at least ’96, arguing about the name. ‘Intelligent Systems for Molecular Biology’ came about because we thought at the time that it was going to be artificial intelligence for the software piece and laboratory robotics was going to be a big piece, instrumentation. Intelligent systems was supposed to include robotics.

---
We had all these grand ideas -- it was sort of [based on] how the genome project was going to advance, there would be both instruments and computer science. So that’s where the Intelligent Systems piece came from—it was supposed to be not just AI, but also robotics.

- Lawrence Hunter ([10.1371/journal.pcbi.1002679](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3431341/))


---
<img
	src="images/introduction/compbio1994.png"
	height="600"
	style="background:white; padding: 0px; margin: 0px"><br>



---
# 2004 

Bioconductor project is launched.


---
## What is bioinformatics?

Early applications include:

- Structural biology
- Phylogenetics
- Data management and databases
- Systems analysis
- Software
- Simulation
- Artificial Intelligence
- Genome sequences

---
Its history is intertwined with:

<div class="fragment">Data</div>
<div class="fragment">Genome sequences</div>
<div class="fragment">Computer hardware</div>

---
### History of genome sequencing

- Discovery of structure of DNA (1953)
- Sanger sequencing (1977-1980)
- The [Human Genome Project](https://www.genome.gov/human-genome-project) (1990-2003)
- The [ENCODE project](https://www.genome.gov/Funded-Programs-Projects/ENCODE-Project-ENCyclopedia-Of-DNA-Elements) (2003-2021)
- Sequencing by synthesis (2008-2012)

---
## Definitions

> the study of information processing in biological systems. Hesper and Hogeweg (1970)


---
Altman 1998  
<img
	src="images/introduction/altman1998.png"
	height="600"
	style="background:white; padding: 0px; margin: 0px"><br>

---
NIH 2000
<img
	src="images/introduction/nih2000.png"
	maxwidth="1100"
	style="background:white; padding: 0px; margin: 0px"><br>

---
Luscombe et al. 2001

> Bioinformatics is conceptualizing biology in terms of macromolecules (in the sense of physical-chemistry) and then applying “informatics” techniques (derived from disciplines such as applied maths, computer science, and statistics) to understand and organize the information associated with these molecules, on a large-scale. 

---
Moore 2007  
<img
	src="images/introduction/moore2007.png"
	height="600"
	style="background:white; padding: 0px; margin: 0px"><br>


---
## Introduction to computational biology

What is computational biology?

bioinformatics?



---
## What is Bioconductor?

Source: [software carpentry](https://github.com/carpentries-incubator/bioc-project
)

1. A repository of R packages for biology
2. A core project with core maintainers
3. A community


---
## What is Bioconductor?

- Started in 2001
- NIH-funded
- Core packages maintained by the project
- Contributed packages maintained by the community
- 6-month release cycle

---
## Package types

- *Software* packages are the heart of Bioconductor
- *Annotation* data packages provide self-contained databases of diverse genomic annotations 
- *Experiment* data packages provide self-contained datasets that are often used by software
- *Workflow* packages collect vignettes that demonstrate the combined analysis, but do not provide any new source code or functionality themselves.

---
## Dig in!

1. How many packages does the current release of the Bioconductor project include?

2. How many packages of each type does this number include?

---

## What are biocViews?

- a set of terms from a controlled vocabulary
- used to classify Bioconductor packages
- can help you find useful packages for your use case

---
## Dig in!

In the RNASeq category, two very popular packages are `DESeq2` and `edgeR`. Which one is more popular in terms of download statistics (i.e., lower rank)?

---

## Core packages

The [core team](https://www.bioconductor.org/about/core-team/) maintains the project.  
A project core provides:

- unified build system
- cohesion of releases
- software review
- interoperability of packages
- training materials

---

## Community Resources

- Course materials: <https://www.bioconductor.org/help/>
- Support site: <https://support.bioconductor.org/>
- Mailing list
- Slack channel

---

## Installing Bioconductor packages

```R
install.packages("BiocManager")
BiocManager::install("GenomicRanges")
```

This can also handle CRAN packages

---


## Core package: Biostrings

```R
BiocManager::install("Biostrings")
library("Biostrings")
DNAString("ATCG")

dna1 <- DNAString("ATCGCTTTGA")
matchPattern("GM", dna1, fixed = TRUE)
matchPattern("GM", dna1, fixed = FALSE)  # allow ambiguity
```

---

```R
adapters = readDNAStringSet(filepath = "TruSeq3-PE-2.fa")
```

```R
DNAStringSet object of length 6:
    width seq                                               names               
[1]    34 TACACTCTTTCCCTACACGACGCTCTTCCGATCT                PrefixPE/1
[2]    34 GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT                PrefixPE/2
[3]    34 TACACTCTTTCCCTACACGACGCTCTTCCGATCT                PE1
[4]    34 AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA                PE1_rc
[5]    34 GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT                PE2
[6]    34 AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC                PE2_rc
```
---
## Documentation 

Find documentation, vignettes, etc. on the website:

<https://bioconductor.org/packages/release/bioc/html/Biostrings.html>


---
## Core package: GenomicRanges

```R
BiocManager::install("GenomicRanges")
```

GRanges stores a set of intervals plus sequence name.

IRanges stores a set of intervals over a range of integer coordinates, without sequence names. 

So: A GRanges object is an IRanges object plus a vector of sequence names.

---
IRanges stores ranges: 

```R
demo_iranges <- IRanges(start = c(10, 15), width = c(10, 5))
demo_iranges
```

```R
IRanges object with 2 ranges and 0 metadata columns:
          start       end     width
      <integer> <integer> <integer>
  [1]        10        19        10
  [2]        15        19         5
```

---
Extract information: 

- `start(demo_iranges)` &rarr; `[1] 10 15`
- `end(demo_iranges)` &rarr; `[1] 19 19`
- `width(demo_iranges)` &rarr; `[1] 10  5`

---
Subset:

```R
demo_iranges[1]
```

```R
IRanges object with 1 range and 0 metadata columns:
          start       end     width
      <integer> <integer> <integer>
  [1]        10        19        10
```

---
Find overlaps:

```R
query_iranges <- IRanges(
  start = c(8, 16),
  end   = c(14, 18)
)
overlaps_iranges <- findOverlaps(query = query_iranges,
								 subject = demo_iranges)
overlaps_iranges
```

```R
Hits object with 3 hits and 0 metadata columns:
      queryHits subjectHits
      <integer>   <integer>
  [1]         1           1
  [2]         2           1
  [3]         2           2
  -------
  queryLength: 2 / subjectLength: 2
```

---

GRanges adds `seqnames`:

```R
demo_granges <- GRanges(
  seqnames = c("chr1", "chr2"),
  ranges = IRanges(
    start = c(10, 20),
    end   = c(25, 35))
)
demo_granges
```

```R
GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         <Rle> <IRanges>  <Rle>
  [1]     chr1     10-25      *
  [2]     chr2     20-35      *
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths
```

---

You can add metadata columns:

```R
GRanges object with 5 ranges and 7 metadata columns:
      seqnames          ranges strand |      source       type     score
         <Rle>       <IRanges>  <Rle> |    <factor>   <factor> <numeric>
  [1]        7 5527147-5529949      - | rtracklayer transcript        NA
  [2]        7 5527147-5530581      - | rtracklayer transcript        NA
  [3]        7 5527147-5530604      - | rtracklayer transcript        NA
  [4]        7 5527147-5530604      - | rtracklayer transcript        NA
  [5]        7 5527147-5530604      - | rtracklayer transcript        NA
          phase         gene_id   gene_name   transcript_id
      <integer>     <character> <character>     <character>
  [1]      <NA> ENSG00000075624        ACTB ENST00000642480
  [2]      <NA> ENSG00000075624        ACTB ENST00000676397
  [3]      <NA> ENSG00000075624        ACTB ENST00000676319
  [4]      <NA> ENSG00000075624        ACTB ENST00000676189
  [5]      <NA> ENSG00000075624        ACTB ENST00000473257
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
---
GRangesList objects hold multiple GRanges

```R
GRangesList object of length 23:
$ENST00000414620
GRanges object with 4 ranges and 7 metadata columns:
      seqnames          ranges strand |      source     type     score
         <Rle>       <IRanges>  <Rle> |    <factor> <factor> <numeric>
  [1]        7 5562574-5562790      - | rtracklayer     exon        NA
  [2]        7 5540676-5540771      - | rtracklayer     exon        NA
  [3]        7 5529535-5529663      - | rtracklayer     exon        NA
  [4]        7 5529282-5529400      - | rtracklayer     exon        NA
          phase         gene_id   gene_name   transcript_id
      <integer>     <character> <character>     <character>
  [1]      <NA> ENSG00000075624        ACTB ENST00000414620
  [2]      <NA> ENSG00000075624        ACTB ENST00000414620
  [3]      <NA> ENSG00000075624        ACTB ENST00000414620
  [4]      <NA> ENSG00000075624        ACTB ENST00000414620
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths

$ENST00000417101
GRanges object with 3 ranges and 7 metadata columns:
      seqnames          ranges strand |      source     type     score
         <Rle>       <IRanges>  <Rle> |    <factor> <factor> <numeric>
  [1]        7 5529806-5529982      - | rtracklayer     exon        NA
  [2]        7 5529535-5529663      - | rtracklayer     exon        NA
  [3]        7 5529235-5529400      - | rtracklayer     exon        NA
          phase         gene_id   gene_name   transcript_id
      <integer>     <character> <character>     <character>
  [1]      <NA> ENSG00000075624        ACTB ENST00000417101
  [2]      <NA> ENSG00000075624        ACTB ENST00000417101
  [3]      <NA> ENSG00000075624        ACTB ENST00000417101
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths

$ENST00000425660
GRanges object with 7 ranges and 7 metadata columns:
      seqnames          ranges strand |      source     type     score
         <Rle>       <IRanges>  <Rle> |    <factor> <factor> <numeric>
  [1]        7 5530524-5530601      - | rtracklayer     exon        NA
  [2]        7 5529535-5529663      - | rtracklayer     exon        NA
  [3]        7 5529161-5529400      - | rtracklayer     exon        NA
  [4]        7 5529019-5529059      - | rtracklayer     exon        NA
  [5]        7 5528281-5528719      - | rtracklayer     exon        NA
  [6]        7 5528004-5528185      - | rtracklayer     exon        NA
  [7]        7 5527156-5527891      - | rtracklayer     exon        NA
          phase         gene_id   gene_name   transcript_id
      <integer>     <character> <character>     <character>
  [1]      <NA> ENSG00000075624        ACTB ENST00000425660
  [2]      <NA> ENSG00000075624        ACTB ENST00000425660
  [3]      <NA> ENSG00000075624        ACTB ENST00000425660
  [4]      <NA> ENSG00000075624        ACTB ENST00000425660
  [5]      <NA> ENSG00000075624        ACTB ENST00000425660
  [6]      <NA> ENSG00000075624        ACTB ENST00000425660
  [7]      <NA> ENSG00000075624        ACTB ENST00000425660
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths

...
<20 more elements>
```

---
## Summary

- GenomicRanges is a useful core package for genomic interval analysis.
- Biostrings is a useful core package for sequence analysis


---
## Other popular packages

- SummarizedExperiment - a data structure for one or more assays, each represented by a matrix-like object of numeric or other mode. The rows typically represent genomic ranges of interest and the columns represent samples. 
- limma - Data analysis, linear models and differential expression for microarray data. 
- BiocFileCache - creates a persistent on-disk cache of files for managing resources that are costly or difficult to create.

---
## Other popular packages

- GenomicAlignments - efficient containers for storing and manipulating short genomic alignments (
clusterProfiler - 
- SingleCellExperiment - class for storing data from single-cell experiments
- ComplexHeatmap - provides a highly flexible way to arrange multiple heatmaps and supports various annotation graphics

---
### Conclusion

It's an exciting time for computational biology.<br>

<div class="fragment">
	The world is starving for people who can apply advanced computation to growing biological data.
</div><br>


<div class="fragment">
	Looking forward to a fun and enlightening course.
</div>

